lib/logstash/outputs/datadog_logs.rb

```diff
@@ -6,67 +6,193 @@
+      batches = batch_events(events, DD_MAX_BATCH_SIZE)
+      batches.each do |batched_event|
+        @codec.encode(batched_event)
+      end
+    else
+      events.each do |event|
+        @codec.encode(event)
+      end
+    end
+  end
+
+  # Encode payload for Datadog to the right format (no-op for HTTP)
+  def encode(payload, use_http, api_key)
+    if not use_http
+      return "#{api_key} #{payload}"
+    else
+      return payload
+    end
+  end
+
+  # Group events in batches
+  def batch_events(events, max_batch_size)
+    batches = []
+    current_batch = []
+    events.each_with_index do |event, i|
+      if i > 0 and i % max_batch_size == 0
+        batches << current_batch
+        current_batch = []
+      end
+      current_batch << event
+    end
+    batches << current_batch
+    return batches
+  end
+
+  # Compress logs with GZIP
+  def gzip_compress(payload, compression_level)
+    gz = StringIO.new
+    gz.set_encoding("BINARY")
+    z = Zlib::GzipWriter.new(gz, compression_level)
+    z.write(payload)
+    z.close
+    gz.string
+  end
+
+  # Build a new transport client
+  def new_client(logger, api_key, use_http, use_ssl, no_ssl_validation, host, port, use_compression)
+    if use_http
+      return DatadogHTTPClient.new logger, use_ssl, no_ssl_validation, host, port, use_compression, api_key
+    else
+      return DatadogTCPClient.new logger, use_ssl, no_ssl_validation, host, port
+    end
+  end
+
+  class RetryableError < StandardError;
+  end
+
+  class DatadogClient
+    def send_retries(payload, max_retries, max_backoff)
       backoff = 1
+      retries = 0
       begin
-        client ||= new_client
-        client.write(message)
-      rescue => e
-        @logger.warn("Could not send payload", :exception => e, :backtrace => e.backtrace)
-        client.close rescue nil
-        client = nil
+        send(payload)
+      rescue RetryableError => e
         if retries < max_retries || max_retries < 0
+          @logger.warn("Retrying ", :exception => e, :backtrace => e.backtrace)
           sleep backoff
           backoff = 2 * backoff unless backoff > max_backoff
           retries += 1
           retry
         end
-        @logger.warn("Max number of retries reached, dropping the payload", :payload => payload, :max_retries => max_retries)
       end
     end
+
+    def send(payload)
+    end
   end
 
-  public
-  def receive(event)
-    # handle new event
-    @codec.encode(event)
+  class DatadogHTTPClient < DatadogClient
+    require "manticore"
+
+    def initialize(logger, use_ssl, no_ssl_validation, host, port, use_compression, api_key)
+      @logger = logger
+      protocol = use_ssl ? "https" : "http"
+      @url = "#{protocol}://#{host}:#{port.to_s}/v1/input/#{api_key}"
+      @headers = {"Content-Type" => "application/json"}
+      if use_compression
+        @headers["Content-Encoding"] = "gzip"
+      end
+      logger.info("Starting HTTP connection to #{protocol}://#{@host}:#{port.to_s}")
```

Comment:
You should probably also log the use or not of the compression

Reply:
Added